File: backend/src/main/java/com/leavemarker/service/EmployeeService.java
package com.leavemarker.service;

import com.leavemarker.dto.employee.EmployeeRequest;
import com.leavemarker.dto.employee.EmployeeResponse;
import com.leavemarker.dto.employee.EmployeeUpdateRequest;
import com.leavemarker.entity.Company;
import com.leavemarker.entity.Employee;
import com.leavemarker.enums.EmployeeStatus;
import com.leavemarker.exception.BadRequestException;
import com.leavemarker.exception.ResourceNotFoundException;
import com.leavemarker.repository.CompanyRepository;
import com.leavemarker.repository.EmployeeRepository;
import com.leavemarker.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class EmployeeService {

    private final EmployeeRepository employeeRepository;
    private final CompanyRepository companyRepository;
    private final PasswordEncoder passwordEncoder;
    private final SubscriptionFeatureService subscriptionFeatureService;

    @Transactional
    public EmployeeResponse createEmployee(EmployeeRequest request, UserPrincipal currentUser) {
        Company company = companyRepository.findById(currentUser.getCompanyId())
                .orElseThrow(() -> new ResourceNotFoundException("Company not found"));

        // Check subscription plan limits
        subscriptionFeatureService.validateCanAddEmployee(company.getId());

        if (employeeRepository.existsByCompanyIdAndEmailAndDeletedFalse(company.getId(), request.getEmail())) {
            throw new BadRequestException("Employee with this email already exists");
        }

        if (employeeRepository.existsByCompanyIdAndEmployeeIdAndDeletedFalse(company.getId(), request.getEmployeeId())) {
            throw new BadRequestException("Employee ID already exists");
        }

        Employee manager = null;
        if (request.getManagerId() != null) {
            manager = employeeRepository.findByIdAndDeletedFalse(request.getManagerId())
                    .orElseThrow(() -> new ResourceNotFoundException("Manager not found"));
        }

        Employee employee = Employee.builder()
                .company(company)
                .employeeId(request.getEmployeeId())
                .fullName(request.getFullName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(request.getRole())
                .department(request.getDepartment())
                .jobTitle(request.getJobTitle())
                .dateOfJoining(request.getDateOfJoining())
                .employmentType(request.getEmploymentType())
                .workLocation(request.getWorkLocation())
                .manager(manager)
                .status(EmployeeStatus.ACTIVE)
                .build();

        employee = employeeRepository.save(employee);
        return mapToResponse(employee);
    }

    public EmployeeResponse getEmployee(Long id, UserPrincipal currentUser) {
        Employee employee = employeeRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found"));

        if (!employee.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        return mapToResponse(employee);
    }

    public List<EmployeeResponse> getAllEmployees(UserPrincipal currentUser) {
        List<Employee> employees = employeeRepository.findByCompanyIdAndDeletedFalse(currentUser.getCompanyId());
        return employees.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public List<EmployeeResponse> getActiveEmployees(UserPrincipal currentUser) {
        List<Employee> employees = employeeRepository.findByCompanyIdAndStatusAndDeletedFalse(
                currentUser.getCompanyId(), EmployeeStatus.ACTIVE);
        return employees.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public EmployeeResponse updateEmployee(Long id, EmployeeUpdateRequest request, UserPrincipal currentUser) {
        Employee employee = employeeRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found"));

        if (!employee.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        if (request.getFullName() != null) {
            employee.setFullName(request.getFullName());
        }
        if (request.getEmail() != null && !request.getEmail().equals(employee.getEmail())) {
            if (employeeRepository.existsByCompanyIdAndEmailAndDeletedFalse(
                    currentUser.getCompanyId(), request.getEmail())) {
                throw new BadRequestException("Employee with this email already exists");
            }
            employee.setEmail(request.getEmail());
        }
        if (request.getRole() != null) {
            employee.setRole(request.getRole());
        }
        if (request.getDepartment() != null) {
            employee.setDepartment(request.getDepartment());
        }
        if (request.getJobTitle() != null) {
            employee.setJobTitle(request.getJobTitle());
        }
        if (request.getDateOfJoining() != null) {
            employee.setDateOfJoining(request.getDateOfJoining());
        }
        if (request.getEmploymentType() != null) {
            employee.setEmploymentType(request.getEmploymentType());
        }
        if (request.getWorkLocation() != null) {
            employee.setWorkLocation(request.getWorkLocation());
        }
        if (request.getManagerId() != null) {
            Employee manager = employeeRepository.findByIdAndDeletedFalse(request.getManagerId())
                    .orElseThrow(() -> new ResourceNotFoundException("Manager not found"));
            employee.setManager(manager);
        }
        if (request.getStatus() != null) {
            employee.setStatus(request.getStatus());
        }

        employee = employeeRepository.save(employee);
        return mapToResponse(employee);
    }

    @Transactional
    public void deactivateEmployee(Long id, UserPrincipal currentUser) {
        Employee employee = employeeRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found"));

        if (!employee.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        employee.setDeleted(true);
        employeeRepository.save(employee);
    }

    private EmployeeResponse mapToResponse(Employee employee) {
        return EmployeeResponse.builder()
                .id(employee.getId())
                .employeeId(employee.getEmployeeId())
                .fullName(employee.getFullName())
                .email(employee.getEmail())
                .role(employee.getRole())
                .department(employee.getDepartment())
                .jobTitle(employee.getJobTitle())
                .dateOfJoining(employee.getDateOfJoining())
                .employmentType(employee.getEmploymentType())
                .workLocation(employee.getWorkLocation())
                .status(employee.getStatus())
                .managerId(employee.getManager() != null ? employee.getManager().getId() : null)
                .managerName(employee.getManager() != null ? employee.getManager().getFullName() : null)
                .companyId(employee.getCompany().getId())
                .companyName(employee.getCompany().getName())
                .build();
    }
}

File: backend-go/internal/service/employee_service.go
package service

import (
	"github.com/leavemarker/backend-go/internal/dto"
	"github.com/leavemarker/backend-go/internal/enums"
	"github.com/leavemarker/backend-go/internal/models"
	"github.com/leavemarker/backend-go/internal/repository"
	"github.com/leavemarker/backend-go/internal/utils"
)

// EmployeeService handles employee business logic
type EmployeeService struct {
	employeeRepo          *repository.EmployeeRepository
	planValidationService *PlanValidationService
}

// NewEmployeeService creates a new EmployeeService
func NewEmployeeService(
	employeeRepo *repository.EmployeeRepository,
	planValidationService *PlanValidationService,
) *EmployeeService {
	return &EmployeeService{
		employeeRepo:          employeeRepo,
		planValidationService: planValidationService,
	}
}

// CreateEmployee creates a new employee
func (s *EmployeeService) CreateEmployee(companyID uint, req dto.EmployeeRequest) (*dto.EmployeeResponse, error) {
	// Validate plan limit
	if err := s.planValidationService.ValidateEmployeeLimit(companyID); err != nil {
		return nil, err
	}

	// Check if email exists
	exists, err := s.employeeRepo.ExistsByEmail(req.Email)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to check email")
	}
	if exists {
		return nil, utils.NewBadRequestError("Email already exists")
	}

	// Check if employee ID exists in company
	existing, err := s.employeeRepo.FindByCompanyIDAndEmployeeID(companyID, req.EmployeeID)
	if err == nil && existing != nil {
		return nil, utils.NewBadRequestError("Employee ID already exists in company")
	}

	// Hash password
	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to hash password")
	}

	employee := &models.Employee{
		CompanyID:      companyID,
		EmployeeID:     req.EmployeeID,
		FullName:       req.FullName,
		Email:          req.Email,
		Password:       hashedPassword,
		Role:           req.Role,
		Department:     req.Department,
		JobTitle:       req.JobTitle,
		DateOfJoining:  req.DateOfJoining,
		EmploymentType: req.EmploymentType,
		WorkLocation:   req.WorkLocation,
		ManagerID:      req.ManagerID,
		Status:         enums.EmployeeStatusActive,
	}

	if err := s.employeeRepo.Create(employee); err != nil {
		return nil, utils.NewInternalServerError("Failed to create employee")
	}

	return s.toEmployeeResponse(employee), nil
}

// GetEmployee gets an employee by ID
func (s *EmployeeService) GetEmployee(id, companyID uint) (*dto.EmployeeResponse, error) {
	employee, err := s.employeeRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Employee not found")
	}

	if employee.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	return s.toEmployeeResponse(employee), nil
}

// GetAllEmployees gets all employees for a company
func (s *EmployeeService) GetAllEmployees(companyID uint) ([]dto.EmployeeResponse, error) {
	employees, err := s.employeeRepo.FindByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get employees")
	}

	responses := make([]dto.EmployeeResponse, len(employees))
	for i, emp := range employees {
		responses[i] = *s.toEmployeeResponse(&emp)
	}

	return responses, nil
}

// GetActiveEmployees gets all active employees for a company
func (s *EmployeeService) GetActiveEmployees(companyID uint) ([]dto.EmployeeResponse, error) {
	employees, err := s.employeeRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get employees")
	}

	responses := make([]dto.EmployeeResponse, len(employees))
	for i, emp := range employees {
		responses[i] = *s.toEmployeeResponse(&emp)
	}

	return responses, nil
}

// CountActiveEmployees counts active employees for a company
func (s *EmployeeService) CountActiveEmployees(companyID uint) (int64, error) {
	return s.employeeRepo.CountActiveByCompanyID(companyID)
}

// UpdateEmployee updates an employee
func (s *EmployeeService) UpdateEmployee(id, companyID uint, req dto.EmployeeUpdateRequest) (*dto.EmployeeResponse, error) {
	employee, err := s.employeeRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Employee not found")
	}

	if employee.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	// Check if email is being changed and if new email exists
	if req.Email != "" && req.Email != employee.Email {
		exists, err := s.employeeRepo.ExistsByEmail(req.Email)
		if err != nil {
			return nil, utils.NewInternalServerError("Failed to check email")
		}
		if exists {
			return nil, utils.NewBadRequestError("Email already exists")
		}
		employee.Email = req.Email
	}

	if req.EmployeeID != "" {
		employee.EmployeeID = req.EmployeeID
	}
	if req.FullName != "" {
		employee.FullName = req.FullName
	}
	if req.Role != "" {
		employee.Role = req.Role
	}
	if req.Department != "" {
		employee.Department = req.Department
	}
	if req.JobTitle != "" {
		employee.JobTitle = req.JobTitle
	}
	if req.DateOfJoining != nil {
		employee.DateOfJoining = req.DateOfJoining
	}
	if req.EmploymentType != "" {
		employee.EmploymentType = req.EmploymentType
	}
	if req.WorkLocation != "" {
		employee.WorkLocation = req.WorkLocation
	}
	employee.ManagerID = req.ManagerID

	if err := s.employeeRepo.Update(employee); err != nil {
		return nil, utils.NewInternalServerError("Failed to update employee")
	}

	return s.toEmployeeResponse(employee), nil
}

// DeactivateEmployee deactivates an employee
func (s *EmployeeService) DeactivateEmployee(id, companyID uint) error {
	employee, err := s.employeeRepo.FindByID(id)
	if err != nil {
		return utils.NewNotFoundError("Employee not found")
	}

	if employee.CompanyID != companyID {
		return utils.NewForbiddenError("Access denied")
	}

	employee.Status = enums.EmployeeStatusInactive
	return s.employeeRepo.Update(employee)
}

// ReactivateEmployee reactivates an employee
func (s *EmployeeService) ReactivateEmployee(id, companyID uint) (*dto.EmployeeResponse, error) {
	employee, err := s.employeeRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Employee not found")
	}

	if employee.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	// Validate plan limit before reactivating
	if err := s.planValidationService.ValidateEmployeeLimit(companyID); err != nil {
		return nil, err
	}

	employee.Status = enums.EmployeeStatusActive
	if err := s.employeeRepo.Update(employee); err != nil {
		return nil, utils.NewInternalServerError("Failed to reactivate employee")
	}

	return s.toEmployeeResponse(employee), nil
}

func (s *EmployeeService) toEmployeeResponse(emp *models.Employee) *dto.EmployeeResponse {
	response := &dto.EmployeeResponse{
		ID:             emp.ID,
		EmployeeID:     emp.EmployeeID,
		FullName:       emp.FullName,
		Email:          emp.Email,
		Role:           emp.Role,
		Department:     emp.Department,
		JobTitle:       emp.JobTitle,
		DateOfJoining:  emp.DateOfJoining,
		EmploymentType: emp.EmploymentType,
		WorkLocation:   emp.WorkLocation,
		Status:         emp.Status,
		ManagerID:      emp.ManagerID,
		CompanyID:      emp.CompanyID,
		CreatedAt:      emp.CreatedAt,
		UpdatedAt:      emp.UpdatedAt,
	}

	if emp.Manager != nil {
		response.ManagerName = emp.Manager.FullName
	}

	if emp.Company.ID != 0 {
		response.CompanyName = emp.Company.Name
	}

	return response
}
